File: src/services/projectService.ts
// src/services/projectService.ts

import { apiClient } from '@/services/apiClient';
import { API_PATH } from '@/constants/api-path';
import type {
  CreateProjectPayload,
  ProjectDTO,
  GetProjectsParams,
  PaginatedProjectResponse,
  InviteMemberPayload,
  InviteMemberResponse,
  ProjectDetailDTO,
  DeleteMemberParams,
  DeleteProjectParams,
} from '@/types/project';
import type { ArchitectureData } from '@/types/architecture';
import type { ComponentListData } from '@/types/component';
import { useProjectStore } from '@/stores/projectStore';

// 프로젝트 생성 함수
export const createProject = async (
  payload: CreateProjectPayload,
): Promise<ProjectDTO> => {
  try {
    const newProject = await apiClient.post<ProjectDTO>(
      API_PATH.PROJECT,
      payload,
    );

    useProjectStore.getState().addProject(newProject);

    return newProject;
  } catch (error) {
    console.error('프로젝트 생성 실패 ', error);
    throw error;
  }
};

/**
 * 프로젝트 목록 조회 함수 (GET /api/projects)
 * @param params - { page, size, sort, order }
 */
export const fetchProjects = async (
  params?: GetProjectsParams,
): Promise<PaginatedProjectResponse> => {
  try {
    const response = await apiClient.get<PaginatedProjectResponse>(
      API_PATH.PROJECT,
      params,
    );

    return response;
  } catch (error) {
    console.error('프로젝트 목록 조회 실패 ', error);
    throw error;
  }
};

/**
 * 프로젝트에 멤버 초대 (POST /api/projects/{projectId}/members)
 * @param projectId - 초대할 프로젝트 ID
 * @param payload - { userId }
 */
export const inviteMember = async (
  projectUuid: string,
  payload: InviteMemberPayload,
): Promise<InviteMemberResponse> => {
  try {
    const response = await apiClient.post<InviteMemberResponse>(
      API_PATH.INVITE_MEMBERS(projectUuid),
      payload,
    );

    useProjectStore.getState().incrementMemberCount(projectUuid);

    return response;
  } catch (error) {
    console.error('멤버 초대 실패', error);
    throw error;
  }
};

// 프로젝트 상세 조회 함수
/**
 * 프로젝트 상세 조회 (GET /api/projects/{projectId})
 * @param projectId - 조회할 프로젝트 ID
 */
export const getProjectDetail = async (
  projectUuid: string,
): Promise<ProjectDetailDTO> => {
  try {
    const projectDetail = await apiClient.get<ProjectDetailDTO>(
      API_PATH.PROJECT_DETAIL(projectUuid),
    );

    useProjectStore.getState().setCurrentProject(projectDetail);

    return projectDetail;
  } catch (error) {
    console.error('프로젝트 상세 조회 실패', error);
    throw error;
  }
};

/**
 * 프로젝트 삭제 (DELETE /api/projects/{projectId})
 * @param projectId - 삭제할 프로젝트 ID
 */
export const deleteProject = async ({
  projectUuid,
}: DeleteProjectParams): Promise<void> => {
  try {
    await apiClient.delete<void>(API_PATH.DELETE_PROJECT(projectUuid));

    useProjectStore.getState().removeProject(projectUuid);
  } catch (error) {
    console.error('프로젝트 삭제 실패', error);
    throw error;
  }
};

// 프로젝트 내 멤버 삭제(DELETE /api/projects/{projectId}/members/{memberId})
export const deleteMember = async ({
  projectUuid,
  memberId,
}: DeleteMemberParams): Promise<void> => {
  try {
    await apiClient.delete<void>(API_PATH.DELETE_MEMBER(projectUuid, memberId));

    useProjectStore.getState().removeMember(projectUuid, memberId);
  } catch (error) {
    console.log('멤버 삭제 실패', error);
    throw error;
  }
};

/**
 * 프로젝트 아키텍처 흐름 조회 (GET /api/projects/{projectId}/architecture)
 * @param projectId - 조회할 프로젝트 ID
 * @param params - { startDate, endDate } 기간 파라미터 (옵션)
 */
export const getArchitecture = async (
  projectId: number,
  params?: { startDate?: string; endDate?: string },
): Promise<ArchitectureData> => {
  try {
    const architecture = await apiClient.get<ArchitectureData>(
      API_PATH.ARCHITECTURE(String(projectId)),
      params,
    );

    return architecture;
  } catch (error) {
    console.error('아키텍처 조회 실패', error);
    throw error;
  }
};

/**
 * 프로젝트 컴포넌트 목록 조회 (GET /api/projects/{projectId}/components)
 * @param projectId - 조회할 프로젝트 ID
 * @param params - { limit, offset } 페이지네이션 파라미터 (옵션)
 */
export const getComponents = async (
  projectId: number,
  params?: { limit?: number; offset?: number },
): Promise<ComponentListData> => {
  try {
    const components = await apiClient.get<ComponentListData>(
      API_PATH.COMPONENTS(String(projectId)),
      params,
    );

    return components;
  } catch (error) {
    console.error('컴포넌트 목록 조회 실패', error);
    throw error;
  }
};

File: src/main/java/S13P31A306/loglens/domain/project/service/ProjectService.java
package S13P31A306.loglens.domain.project.service;

import S13P31A306.loglens.domain.project.dto.request.ProjectCreateRequest;
import S13P31A306.loglens.domain.project.dto.request.ProjectMemberInviteRequest;
import S13P31A306.loglens.domain.project.dto.response.ProjectCreateResponse;
import S13P31A306.loglens.domain.project.dto.response.ProjectDetailResponse;
import S13P31A306.loglens.domain.project.dto.response.ProjectListResponse;
import S13P31A306.loglens.domain.project.dto.response.ProjectMemberInviteResponse;

/**
 * 프로젝트 관련 비즈니스 로직을 처리하는 서비스
 */
public interface ProjectService {

    /**
     * 프로젝트 생성
     *
     * @param request 생성할 프로젝트 정보
     * @return ProjectCreateResponse 생성한 프로젝트 정보
     */
    ProjectCreateResponse createProject(ProjectCreateRequest request);

    /**
     * 멤버 초대
     *
     * @param projectUuid 사용자를 초대할 프로젝트 UUID
     * @param request     초대할 사용자 ID
     * @return ProjectMemberInviteResponse 초대한 사용자 정보
     */
    ProjectMemberInviteResponse inviteMember(String projectUuid, ProjectMemberInviteRequest request);

    /**
     * 프로젝트 목록 조회
     *
     * @param page  출력할 페이지 번호 - default 0
     * @param size  한 페이지에 나타낼 프로젝트의 수 - default 10
     * @param sort  정렬 조건 - default createdAt
     * @param order 정렬 방향 - default desc
     * @return ProjectListResponse 프로젝트 리스트
     */
    ProjectListResponse getProjects(int page, int size, String sort, String order);

    /**
     * 프로젝트 상세 조회
     *
     * @param projectUuid 프로젝트 UUID
     * @return ProjectDetailResponse
     */
    ProjectDetailResponse getProjectDetail(String projectUuid);

    Integer getProjectIdByUuid(String uuid);

    /**
     * 프로젝트 삭제
     *
     * @param projectUuid 프로젝트 UUID
     */
    void deleteProject(String projectUuid);

    /**
     * 멤버 삭제
     *
     * @param projectUuid 프로젝트 UUID
     * @param memberId    멤버 ID
     */
    void deleteMember(String projectUuid, int memberId);
}
